sys/cheri/cheri_colocation.c

```diff
@@ -616,6 +625,17 @@ kern_cosetup(struct thread *td, int what,
 			    td->td_proc->p_sysent->sv_cocall_base,
 			    td->td_proc->p_sysent->sv_cocall_len);
 		}
+#ifdef __CHERI_PURE_CAPABILITY__
+		/*
+		 * The least significant bit of the code capability needs to be set.
+		 * Otherwise, the CPU leaves C64 mode after the BLRS instruction that
+		 * is used to jump into the switcher code. The switcher code provided
+		 * by purecap kernels uses capabilities and so fails if it is not
+		 * executed in C64 mode.
+		 */
+		codecap_addr = cheri_getaddress(codecap);
```

Comment:
1. Use cheri_capmode, it’s target-independent
2. Shouldn’t switcher_code_cap do this?
3. This breaks for the vm_cocall_codecap case, because you add not or (IIRC cheri_capmode is idempotent) and so will double capmode
4. Why is this based on kernel ABI? Surely it should always be purecap code?

Reply:
@jrtc27 I implemented these suggestions in this commit: https://github.com/CTSRD-CHERI/cheribsd/pull/1791/commits/b1ee3480963f3fc6c958a8910eb14b2f954e4730
I found more bugs in the coprocess implementation for Morello in the meantime. I have fixes for these and I'm hoping to find the time to add these to this PR soon.